crates/dojo/types/src/primitive.rs

```diff
@@ -403,6 +403,42 @@ impl Primitive {
             }
         }
     }
+
+    pub fn can_upgrade_to(&self, other: &Primitive) -> bool {
+        use Primitive::*;
+        match (self, other) {
+            // Boolean (no upgrades)
+            (Bool(_), Bool(_)) => true,
+
+            // Unsigned integer upgrades
+            (U8(_), U16(_) | U32(_) | USize(_) | U64(_) | U128(_) | Felt252(_)) => true,
+            (U16(_), U32(_) | USize(_) | U64(_) | U128(_) | Felt252(_)) => true,
+            (U32(_), USize(_) | U64(_) | U128(_) | Felt252(_)) => true,
+            (USize(_), U32(_) | U64(_) | U128(_) | Felt252(_)) => true,
+            (U64(_), U128(_) | Felt252(_)) => true,
+            (U128(_), Felt252(_)) => true,
+
+            // U256 can only upgrade to itself
+            (U256(_), U256(_)) => true,
+
+            // Signed integer upgrades
+            (I8(_), I16(_) | I32(_) | I64(_) | I128(_) | Felt252(_)) => true,
+            (I16(_), I32(_) | I64(_) | I128(_) | Felt252(_)) => true,
+            (I32(_), I64(_) | I128(_) | Felt252(_)) => true,
+            (I64(_), I128(_) | Felt252(_)) => true,
+            (I128(_), Felt252(_)) => true,
+
+            // Felt252 and address types
+            (Felt252(_), ClassHash(_) | ContractAddress(_)) => true,
+            (ClassHash(_), Felt252(_) | ContractAddress(_)) => true,
+            (ContractAddress(_), Felt252(_) | ClassHash(_)) => true,
```

Comment:
We're missing `EthAddress`, not sure at which point it's relevant for now and may be supported later.

Reply:
Hmm we dont even have EthAddress on the Ty in torii